### ISSUE TITLE
Support for playlists

### ISSUE THREAD

**Issue by MartijnBraam** [budkin/jam#10]
Is there a possibility to add playlists?

I'm not 100% sure how it would work since the playlists can contain tracks that are not in your library and fetching all the playlist tracks to jam would in my case add a lot of artists with a single track.

**Comment by budkin** [comment]
Yes! Single track artists is a problem. I have a ton of them. I'm trying to figure out a way how to merge them into "Various Artists" and show the tracks in corresponding albums.

As for playlists.. I'll think about it as well, the gmusic library has methods to deal with playlists, so it's absolutely possible, I don't even need to write additional methods

**Comment by MartijnBraam** [comment]
Maybe have the playlists as alternative view for the artist list and hide artists from playlists from the artist list there is now.

**Comment by budkin** [comment]
I'm thinking about putting playlist into a special artist

**Comment by budkin** [comment]
How do you like my idea? "Playlists" artist

**Comment by MartijnBraam** [comment]
It's an easy way but it's probably better to slightly alter the interface

Change the header for the left column to Artist / Playlist and the right column to Album / Track.
Have the tab key switch between 3 states: the track panel, the artist/playlist panel and the header of the left column. Making the left column have "tabs"

**Comment by budkin** [comment]
I'm planning to add radio stations as well, do you propose a separate tab for them too?

**Comment by MartijnBraam** [comment]
That would be nicely combined as tabs in the first columns. I don't use the radio stations a lot but aren't they dynamically generated? The time I used I saw the tracks to be played change during playing.

**Comment by howardcochran** [comment]
MartijnBraam wrote:
> Change the header for the left column to Artist / Playlist and the right column to Album / Track.

+1
Playlists (particularly the ThumbsUp list) is the most important feature for me, as that is how all my music is organized.  Having it as a separate view would be great. The right pane could just list list artist & track in the order they will be played (i.e. not sorted by artist).  And need to be able to shuffle the list.

'Wish I could help you out with this, but my day job has me pegged right now.  Great work, though!

**Comment by budkin** [comment]
Unfortunately, the ThumbUp playlist is not retrievable, see https://github.com/simon-weber/gmusicapi/issues/411

**Comment by howardcochran** [comment]
> Unfortunately, the ThumbUp playlist is not retrievable, see simon-weber/gmusicapi#411

Well, that is unfortunate.  In that case, support for normal (non-auto) playlists would still be very useful, because I think I can use the gmusic web API to manually copy the contents of the ThumbsUp list to a named, static playlist.  It adds a bit of manual maintenance, but at least it would be possible.

**Comment by budkin** [comment]
I'm working on playlists as I type this, you will be able to toggle playlists/Artists view by pressing ctrl+space. The problem is I don't have a lot of playlists and can't test it thoroughly, so this is going to be an alpha release

**Comment by howardcochran** [comment]
@budkin, Thanks for working on playlists!

**Comment by budkin** [comment]
Sorry to keep you waiting for so long, the new release is ready

**Comment by CyanJDroid** [comment]
@budkin just tried the new release, and while playlists do work (yay!), only a few of my playlists show up (I have _a lot_ of playlists, some of which have _a lot_ of songs in them), and in the playlists that do show up, only some of the songs in said playlists show up. not sure what can be done to fix this, but figured I should let you know the problems I'm having with it.

**Comment by budkin** [comment]
@CyanJDroid uploaded songs will not be displayed, only songs from GPM will. I know it's frustrating, but I can't really do something about it, such is the MobileClient API

**Comment by CyanJDroid** [comment]
@budkin I understand that uploaded songs won't show (which is unfortunate, if only Google would release a official api!), but some of these playlists are 99% non-uploaded songs, and jam only shows a few songs in the playlists ¯\_(ツ)_/¯

**Comment by CyanJDroid** [comment]
@budkin here's an example, my Chillstep playlist has 164 songs in it, yet in jam it only has one song (only a few of the songs in this playlist are uploaded ones):

![gpm](https://cloud.githubusercontent.com/assets/5077450/23588718/37551e02-0187-11e7-8b83-c615287ff287.png)
![jam](https://cloud.githubusercontent.com/assets/5077450/23588720/3ca242cc-0187-11e7-8947-be258b1da1ac.png)

**Comment by budkin** [comment]
hm, that is a problem. First thing I'd do if I were you is remove the database (~/.local/share/jamdb) and start from scratch. Did you try that?

**Comment by budkin** [comment]
there's absolutely no problem with the "Milkshakes and Chill" track, I just added it to my own playlist, and it is shown

**Comment by CyanJDroid** [comment]
@budkin tried removing the database, but to no avail, I still have playlists and songs missing. perhaps it simply can't handle the sheer number of playlists and songs I have?

**Comment by budkin** [comment]
@CyanJDroid that's unlikely. it's a bug somewhere in the code. did you notice if the songs shown are the same or different?

**Comment by CyanJDroid** [comment]
@budkin the same playlists and songs show as before.

**Comment by budkin** [comment]
Okay, I added some logging to the player, try this executable and show me the output (delete the database beforehand)
[a.zip](https://github.com/budkin/jam/files/819446/a.zip)

**Comment by CyanJDroid** [comment]
@budkin this is what I got, that error of "400 Bad Request" repeated _a lot more_ before the app launched

![jam](https://cloud.githubusercontent.com/assets/5077450/23589329/994aa1c8-0190-11e7-9042-141d47dd91dd.png)

**Comment by budkin** [comment]
@CyanJDroid and no other errors? that's strange. error code 400 is supposed to be an uploaded song

**Comment by CyanJDroid** [comment]
@budkin cleared database and ran again just to be sure, and error code 400 is the only thing I get.

**Comment by budkin** [comment]
okay, so I got a playlist that doesn't show up. investigating

**Comment by howardcochran** [comment]
@budkin, thanks for investigating this.  FWIW, my favorite playlist, having 678 songs, only displays 6 songs in jam. It is 100% GPM music. I deleted jamdb before starting.

It ran it using the "a.zip" you uploaded 2 days ago (with extra logging).  It showed "400 Bad Request" 23 times, even though my playlist is missing hundreds of songs.  Dunno if that tidbit helps you.  Lemme know if you'd like me to gather more info.

**Comment by budkin** [comment]
yes, it actually helps me a lot, thank you, it means that processing playlist entries is being done just fine and the problem is somewhere else

**Comment by budkin** [comment]
If someone who experiences this issue wants to help, here's the function that does all job https://github.com/budkin/jam/blob/master/music/lib.go#L72
